File: lib/actions/service/create.js
'use strict';

var inquirer = require('inquirer');
var fs = require('fs');
var sh = require('shelljs');
var path = require('path');
var compose = require('../../compose');
var file = require('../../file');

var handlers = {
  'node.js': node,
  custom: custom
};

module.exports = function create(name) {
  inquirer.prompt([
    {
      type: 'list',
      name: 'template',
      message: 'Which template would you like to use?',
      choices: [
        'custom',
        'node.js'
      ]
    },
    {
      type: 'input',
      name: 'image',
      message: 'What will be the image source for this service?',
      default: path.basename(process.cwd()) + '/' + name + ':latest',
      validate: function(image) {
        return !!image || 'An image source is required';
      }
    },
    {
      type: 'input',
      name: 'base',
      message: 'On what image is this service based on?',
      when: function(answers) {
        return answers.template === 'custom';
      },
      validate: function(base) {
        return !!base || 'A base image is required';
      }
    }
  ], function (answers) {
    var source = __dirname + '/../../../tpl/create/' + answers.template;

    sh.mkdir('-p', name);
    sh.cp(source + '/*', source + '/.*', name);

    file.replace(path.join(name, 'README.md'), {name: name});

    handlers[answers.template](name, answers);

    console.log('New ' + answers.template + ' service create in ' + path.join(process.cwd(), name));
  });
};

function node(name, answers) {
  sh.mkdir('-p', name + '/lib');
  sh.mkdir('-p', name + '/test');
  sh.touch(name + '/lib/.gitkeep');
  sh.touch(name + '/test/.gitkeep');

  compose.addService(name, {
    image: answers.image
  }, {
    build: './' + name,
    command: 'nodemon .',
    volumes: [
      './' + name + '/server.js:/usr/src/app/server.js',
      './' + name + '/lib:/usr/src/app/lib',
      './' + name + '/test:/usr/src/app/test'
    ]
  });
}

function custom(name, answers) {
  fs.writeFileSync(path.join(name, 'Dockerfile'), 'FROM ' + answers.base);

  compose.addService(name, {
    image: answers.image
  }, {
    build: './' + name
  });
}

File: lib/frigate/commands/create.rb
require 'inquirer'
require 'frigate/compose'

module Frigate::Commands
  class Create
    def initialize(args, options)
      @compose = Frigate::Compose.new

      name = args[0]

      template = ask_template
      image = ask_image(name)
      base = nil

      if template == 'custom'
        base = ask_base
      end

      source = File.join(File.join(File.dirname(__FILE__), '/../../../tpl/create'), template)

      FileUtils.mkdir_p(name)
      FileUtils.cp_r(File.join(source, '/.'), name)

      readme = File.join(name, 'README.md')
      File.write(readme, File.read(readme).gsub(/\#\{name\}/, name))

      self.send("handle_#{template}", name, base: base, image: image)

      puts "New #{template} service created in #{File.join(Dir.pwd, name)}"
    end

    private

    def ask_template
      templates = %w(custom node.js)
      values = %w(custom node)
      idx = Ask.list 'Which template would you like to use?', templates
      values[idx]
    end

    def ask_image(name)
      Ask.input 'What will be the image source for this service?',
                default: "#{File.basename(Dir.pwd)}/#{name}:latest"
    end

    def ask_base
      Ask.input 'On what image is this service based on?',
                default: 'scratch'
    end

    def handle_node(name, answers)
      FileUtils.mkdir_p "#{name}/lib"
      FileUtils.mkdir_p "#{name}/test"
      FileUtils.touch("#{name}/lib/.gitkeep")
      FileUtils.touch("#{name}/test/.gitkeep")

      config = {
        'image' => answers[:image]
      }

      override = {
        'build' => "./#{name}",
        'command' => 'nodemon .',
        'volumes' => [
          "./#{name}/server.js:/usr/src/app/server.js",
          "./#{name}/lib:/usr/src/app/lib",
          "./#{name}/test:/usr/src/app/test"
        ]
      }

      @compose.add_service(name, config, override)
    end

    def handle_custom(name, answers)
      File.write(File.join(name, 'Dockerfile'), "FROM #{answers[:base]}")

      @compose.add_service(name, {'image' => answers[:image]}, {'build' => "./#{name}"})
    end
  end
end
